### PR TITLE
feat: implement intrinsic `exit()`

### LINKED ISSUE #4594: Implement exit() intrinsic
This is a GFortran extension, but SNAP uses it. Let's implement this intrinsic, but give a warning that this is a non-standard function, and a tip to use `stop` or `error stop`.

### PR DESCRIPTION
**Pull request by Pranavchiku**
Fixes #4594.

### CODE REVIEW

**Comment by kmr-srbh** [comment]
Thanks for this! I have one thought on this. Should we create the `Stop_t` node for this instead? I think `EXIT` might be closer to it.

### EXIT

```fortran
PROGRAM main
    
    IMPLICIT NONE

    CALL EXIT

END PROGRAM main
```
```console
(base) saurabh-kumar@Awadh:~/Projects/System/lfortran$ gfortran ./examples/example.f90 && ./a.out
(base) saurabh-kumar@Awadh:~/Projects/System/lfortran$
```

### STOP

```fortran
PROGRAM main
    
    IMPLICIT NONE

    STOP

END PROGRAM main
```
```console
(base) saurabh-kumar@Awadh:~/Projects/System/lfortran$ gfortran ./examples/example.f90 && ./a.out
(base) saurabh-kumar@Awadh:~/Projects/System/lfortran$
```

### ERROR STOP

```fortran
PROGRAM main
    
    IMPLICIT NONE

    ERROR STOP

END PROGRAM main
```
```console
(base) saurabh-kumar@Awadh:~/Projects/System/lfortran$ gfortran ./examples/example.f90 && ./a.out
ERROR STOP 

Error termination. Backtrace:
#0  0x7e89f3823c5d in ???
#1  0x7e89f38247b9 in ???
#2  0x7e89f3825e37 in ???
#3  0x57cb8130d170 in ???
#4  0x57cb8130d1a9 in ???
#5  0x7e89f342a1c9 in __libc_start_call_main
        at ../sysdeps/nptl/libc_start_call_main.h:58
#6  0x7e89f342a28a in __libc_start_main_impl
        at ../csu/libc-start.c:360
#7  0x57cb8130d094 in ???
#8  0xffffffffffffffff in ???
```

File: src/libasr/pass/intrinsic_subroutines.h [hunk 1]
```diff
@@ -166,6 +168,31 @@ namespace RandomNumber {
...
+            int kind = ASRUtils::extract_kind_from_ttype_t(arg_type);
+            ASRUtils::require_impl(ASRUtils::is_integer(*arg_type) && (kind == 4), "Overload Id for random_number expected to be 0, found " + std::to_string(x.m_overload_id), x.base.base.loc, diagnostics);
+        }
+    }
+
+    static inline ASR::asr_t* create_Exit(Allocator& al, const Location& loc, Vec<ASR::expr_t*>& args, diag::Diagnostics& /*diag*/) {
+        Vec<ASR::expr_t*> m_args; m_args.reserve(al, 1);
+        ASR::expr_t* arg = args.size() >= 0 ? args[0] : nullptr;
+        return ASR::make_ErrorStop_t(al, loc, arg);
+    }
```
Comment (certik) [inline src/libasr/pass/intrinsic_subroutines.h:185]: Something seems inconsistent here --- isn't the `create` routine supposed to create the intrinsic function, Exit in this case?

If you are just creating ErrorStop, then the frontend should create it directly.

In this case, we can create an Exit intrinsic function, and instantiate it to ErrorStop (or Stop).

**Comment by certik** [comment]
I think ErrorStop always exits with an error. Stop 0 exits normally, and Stop non-zero will exit with an error. So I think Exit corresponds to Stop almost (if not completely) exactly.

The question is whether the frontend shouldn't just insert Stop directly. But we can add an intrinsic subroutine and instantiate it to Stop (see my comment above). It's a bit redundant, but we can clean up redundancies later. It seems like a clean approach.

File: src/libasr/pass/intrinsic_subroutines.h (hunk 1 above)
Comment (Pranavchiku) [reply src/libasr/pass/intrinsic_subroutines.h:185]: I did this with purpose. I do not find any reason to instantiate this, we can directly put ErrorStop or Stop instead. In a few intrinsic elemental function we did instantiation even if the output was fixed to allow broadcasting ( implicitly ). Here there are no such cases. 

Either way is fine, `create_Exit` is called via frontend eventually, so creating from frontend or this way almost does the same thing.

**Comment by Pranavchiku** [comment]
Okay, I'll change it to `Stop`.

File: src/libasr/pass/intrinsic_subroutines.h (hunk 1 above)
Comment (certik) [reply src/libasr/pass/intrinsic_subroutines.h:185]: The problem is that if we don't instantiate, then we don't have this intrinsic Exit function. We only have Stop in ASR. Which is fine, but at that point, let's not add this to our intrinsic function infrastructure at all I think.

This is half-pretending that we have an Exit intrinsic, but we can't actually create (represent!) an Exit intrinsic in ASR.

For example, how is `verify_args` above called? It accepts `ASR::IntrinsicImpureSubroutine_t`, but such an ASR node can't represent Exit. Except it can in:
```
enum class IntrinsicImpureSubroutines : int64_t {
    RandomNumber,
    RandomInit,
    Exit,
    // ...
};
```
But it can't, since you can't create it.

In other words, unless I am missing something, the verify_args is never called, correct? So let's remove it.

And then you are left with just create_Exit. This function should be moved to the frontend.

**Comment by Pranavchiku** [comment]
Continued at #4621
